lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,496 @@
+		l.Warnf(`Folder "%s": Starting FS notifications failed: %s`,
+			fsWatcher.description, err)
+		return nil, err
+	}
+
+	return fsWatcher, nil
+}
+
+func (watcher *fsWatcher) setupNotifications() error {
+	absShouldIgnore := func(absPath string) bool {
+		if !isSubpath(absPath, watcher.folderPath) {
+			return true
+		}
+		relPath, _ := filepath.Rel(watcher.folderPath, absPath)
+		return watcher.ignores.ShouldIgnore(relPath)
+	}
+	// if err := notify.WatchWithFilter(filepath.Join(watcher.folderPath, "..."),
+	if err := notify.WatchWithFilter(filepath.Join(watcher.folderPath, "..."),
+		watcher.fsEventChan, absShouldIgnore, watcher.eventMask()); err != nil {
+		notify.Stop(watcher.fsEventChan)
+		close(watcher.fsEventChan)
+		if isWatchesTooFew(err) {
+			err = watchesLimitTooLowError(watcher.description)
+		}
+		return err
+	}
+	l.Infoln(watcher, "Initiated filesystem watcher")
+	return nil
+}
+
+func (watcher *fsWatcher) Serve() {
+	watcher.notifyTimer = time.NewTimer(watcher.notifyDelay)
+	defer watcher.notifyTimer.Stop()
+
+	inProgressItemSubscription := events.Default.Subscribe(
+		events.ItemStarted | events.ItemFinished)
+
+	for {
+		// Detect channel overflow
+		if len(watcher.fsEventChan) == maxFiles {
+		outer:
+			for {
+				select {
+				case <-watcher.fsEventChan:
+				default:
+					break outer
+				}
+			}
+			// Issue full rescan as events were lost
+			watcher.newFsEvent(watcher.folderPath, nonRemove)
+			l.Debugln(watcher, "Backend channel overflow: Scan entire folder")
+		}
+		select {
+		case event, _ := <-watcher.fsEventChan:
+			watcher.newFsEvent(event.Path(), eventType(event.Event()))
+		case event := <-inProgressItemSubscription.C():
+			watcher.updateInProgressSet(event)
+		case <-watcher.notifyTimer.C:
+			watcher.actOnTimer()
+		case interval := <-watcher.resetNotifyTimerChan:
+			watcher.resetNotifyTimer(interval)
+		case ignores := <-watcher.ignoresUpdate:
+			watcher.ignores = ignores
+		case <-watcher.stop:
+			return
+		}
+	}
+}
+
+func (watcher *fsWatcher) Stop() {
+	close(watcher.stop)
+	notify.Stop(watcher.fsEventChan)
+	l.Infoln(watcher, "Stopped filesystem watcher")
+}
+
+func (watcher *fsWatcher) FsWatchChan() <-chan []string {
+	return watcher.notifyModelChan
+}
+
+func (watcher *fsWatcher) newFsEvent(eventPath string, eventType fsEventType) {
+	if _, ok := watcher.rootEventDir.events["."]; ok {
+		l.Debugf("%v Will scan entire folder anyway; dropping: %s",
+			watcher, eventPath)
+		return
+	}
+	if isSubpath(eventPath, watcher.folderPath) {
+		path, _ := filepath.Rel(watcher.folderPath, eventPath)
+		if watcher.pathInProgress(path) {
+			l.Debugf("%v Skipping notification for path we modified: %s",
+				watcher, path)
+			return
+		}
+		if watcher.ignores.ShouldIgnore(path) {
+			l.Debugf("%v Ignoring: %s", watcher, path)
+			return
+		}
+		watcher.aggregateEvent(path, time.Now(), eventType)
+	} else {
+		l.Warnf("%v Bug: Detected change outside of folder (%v), dropping: %s",
```

Comment:
To be honest, we should probably panic in this case, as something really weird is going on.
But maybe that's too harsh.